texts/0004-wry-raw-window-handle.md

```diff
@@ -0,0 +1,113 @@
+- Start Date: 16-10-2023
+- RFC PR: [tauri-apps/rfcs#12](https://github.com/tauri-apps/rfcs/pull/12)
+- Tracking Issue: [tauri-apps/wry#677](https://github.com/tauri-apps/wry/issues/677)
+
+# Summary
+
+Make WRY agnostic of the windowing library used and accept a raw-window-handle or gtk-window-handle where applicable.
+
+# Motivation
+
+This has been requested in [tauri-apps/wry#677](https://github.com/tauri-apps/wry/issues/677) a long time ago and at work at CrabNebula,
+we have been contracted to research and develop an integration of wry and wgpu/bevy. We had a successful results and it aligned with [tauri-apps/wry#677](https://github.com/tauri-apps/wry/issues/677)
+so we did an initial implemention of this RFC for Windows which can be found here https://github.com/crabnebula-dev/wry/commits/feat/embedded-webview and we think
+it is a good oprtunity to upstream work.
+
+# Guide-level explanation
+There is two groups of window handles we will need to take in order to support the current platforms we support:
+
+#### Raw window handle:
+This group could leverage the `raw-window-handle` crate; this includes macOS, Windows and Linux (x11) platforms.
+  - Using winit, tao, egui or any other windowing library that
+    implements [`raw_window_handle::HasWindowHandle`](https://docs.rs/raw-window-handle/latest/raw_window_handle/trait.HasWindowHandle.html) trait
+    (or previous versions of `raw-window-handle` crate):
+    ```rs
+    let window = Window::new(&evl);
+    let raw_handle = window.raw_window_handle();
+    let webview = WebView::new(raw_handle);  
+    ```
+    
+  - Using raw win32:
+    ```rs
+    let hwnd = CreateWindowExW(/* args */);
+    let raw_handle =  RawWindowHandle::Win32(Win32WindowHandle {
+      hwnd,
+      hinstance: None,
+    });
+    let webview = WebView::new(raw_handle);  
+    ```
+    
+  - Using raw AppKit NSWindow:
+    ```rs
+    let ns_window = msg_send![/* args */];
+    let raw_handle =  RawWindowHandle::Win32(AppKitWindowHandle  {
+      ns_view: ns_window.contentView(),
+    });
+    let webview = WebView::new(raw_handle);  
+    ```
+  - Using raw X11:
+    ```rs
+    let window = XCreateWindow(/* args */);
+    let raw_handle =  RawWindowHandle::Xlib(XlibWindowHandle::new(window);
+    let webview = WebView::new(raw_handle);  
+    ```
+
+#### Gtk handle:
+This group leverages the gtk types, includes Linux platform only (X11 and Wayland). 
+  - Using raw gtk:
+    ```rs
+    let gtk_window = gtk::ApplicationWindow::new(/* args */);
+    let webview = WebView::new_gtk(&gtk_window);  
+    ```
+  - Using tao:
+    ```rs
+    let window = Window::new(/* args */);
+    let gtk_window = window.gtk_window();
+    let webview = WebView::new_gtk(&gtk_window);  
+    ```
+
+
+# Reference-level explanation
+
+There is 2 APIs this RFC is proposing:
+- `Webview::new(parent_raw_window_handle)` initializes the Webview as a child in the passed window handle, this allows for multiple webviews in
```

Comment:
Are we still going to upstream `EmbeddedWebview`? Because this RFC seems more focus on original `Webview` type.
I'm prefer the RFC approach anyway. Looks like a good refactor we should be done for a long time.

Reply:
Indeed, this RFC is aiming to modify the current type without adding a new type.